presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java

```diff
@@ -0,0 +1,237 @@
+        lines = sendElasticQuery(builder).iterator();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return totalBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return columnHandles.get(field).getColumnType();
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        if (!lines.hasNext()) {
+            return false;
+        }
+
+        SearchHit hit = lines.next();
+        fields = new ArrayList(Collections.nCopies(columnHandles.size(), null));
+
+        setFieldIfExists("_id", hit.getId());
+        setFieldIfExists("_index", hit.getIndex());
+
+        extractFromSource(hit);
+        totalBytes += fields.size();
+        return true;
+    }
+
+    @Override
+    public boolean getBoolean(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(BOOLEAN));
+        return (Boolean) getFieldValue(field);
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(BIGINT, INTEGER));
+        return Long.valueOf(String.valueOf(getFieldValue(field)));
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(DOUBLE));
+        return (Double) getFieldValue(field);
+    }
+
+    @Override
+    public Slice getSlice(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(VARCHAR));
+
+        Object value = getFieldValue(field);
+        ElasticsearchColumnHandle column = columnHandles.get(field);
+        if (tupleDomain.getDomains().isPresent()) {
+            Map<ColumnHandle, Domain> domains = tupleDomain.getDomains().get();
+            Domain domain = domains.get(column);
+            if (domain != null) {
+                for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+                    if (range.isSingleValue()) {
+                        return Slices.utf8Slice(String.valueOf(getValue(column.getColumnType(), range.getLow().getValue())));
+                    }
+                }
+            }
+        }
+        if (value instanceof Collection) {
+            return Slices.utf8Slice(new JSONArray((List<Map<String, Object>>) value).toString());
+        }
+        else {
+            return Slices.utf8Slice(String.valueOf(value));
+        }
+    }
+
+    @Override
+    public Object getObject(int field)
+    {
+        return null;
+    }
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return getFieldValue(field) == null;
+    }
+
+    private void checkFieldType(int field, Set<Type> expectedTypes)
+    {
+        Type actual = getType(field);
+        StringBuilder builder = new StringBuilder();
```

Comment:
you don't really need to populate the `builder` if `expectedTypes` contains `actual`.